File: frontend/lib/models/nguoi_dung.dart
class NguoiDung {
  final String id;
  final String hoTen;
  final String email;
  final String vaiTro;
  final String? soDienThoai;
  final String? avatar;
  final String trangThai;
  final DateTime? lanDangNhapCuoi;
  final DateTime createdAt;
  final DateTime updatedAt;

  NguoiDung({
    required this.id,
    required this.hoTen,
    required this.email,
    required this.vaiTro,
    this.soDienThoai,
    this.avatar,
    required this.trangThai,
    this.lanDangNhapCuoi,
    required this.createdAt,
    required this.updatedAt,
  });

  factory NguoiDung.fromJson(Map<String, dynamic> json) {
    return NguoiDung(
      id: json['_id'] ?? '',
      hoTen: json['hoTen'] ?? '',
      email: json['email'] ?? '',
      vaiTro: json['vaiTro'] ?? 'HR',
      soDienThoai: json['soDienThoai'],
      avatar: json['avatar'],
      trangThai: json['trangThai'] ?? 'Hoat_dong',
      lanDangNhapCuoi: json['lanDangNhapCuoi'] != null
          ? DateTime.parse(json['lanDangNhapCuoi'])
          : null,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'hoTen': hoTen,
      'email': email,
      'vaiTro': vaiTro,
      'soDienThoai': soDienThoai,
      'avatar': avatar,
      'trangThai': trangThai,
      'lanDangNhapCuoi': lanDangNhapCuoi?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  NguoiDung copyWith({
    String? id,
    String? hoTen,
    String? email,
    String? vaiTro,
    String? soDienThoai,
    String? avatar,
    String? trangThai,
    DateTime? lanDangNhapCuoi,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return NguoiDung(
      id: id ?? this.id,
      hoTen: hoTen ?? this.hoTen,
      email: email ?? this.email,
      vaiTro: vaiTro ?? this.vaiTro,
      soDienThoai: soDienThoai ?? this.soDienThoai,
      avatar: avatar ?? this.avatar,
      trangThai: trangThai ?? this.trangThai,
      lanDangNhapCuoi: lanDangNhapCuoi ?? this.lanDangNhapCuoi,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  // Helper methods
  bool get isAdmin => vaiTro == 'Admin';
  bool get isHR => vaiTro == 'HR';
  bool get isInterviewer => vaiTro == 'Interviewer';
  bool get isActive => trangThai == 'Hoat_dong';
}

File: backend/src/models/NguoiDung.js
const mongoose = require('mongoose');
const bcrypt = require('bcryptjs');
const { VAI_TRO } = require('../utils/constants');

const nguoiDungSchema = new mongoose.Schema({
  hoTen: {
    type: String,
    required: [true, 'Vui lòng nhập họ tên'],
    trim: true,
    maxlength: [100, 'Họ tên không được quá 100 ký tự']
  },
  email: {
    type: String,
    required: [true, 'Vui lòng nhập email'],
    unique: true,
    lowercase: true,
    trim: true,
    match: [/^\S+@\S+\.\S+$/, 'Email không hợp lệ']
  },
  matKhauHash: {
    type: String,
    required: [true, 'Vui lòng nhập mật khẩu'],
    minlength: [6, 'Mật khẩu phải có ít nhất 6 ký tự'],
    select: false // Không trả về mật khẩu khi query
  },
  vaiTro: {
    type: String,
    enum: Object.values(VAI_TRO),
    default: VAI_TRO.HR,
    required: true
  },
  soDienThoai: {
    type: String,
    trim: true,
    match: [/^[0-9]{10,11}$/, 'Số điện thoại không hợp lệ']
  },
  avatar: {
    type: String,
    default: null
  },
  trangThai: {
    type: String,
    enum: ['Hoat_dong', 'Khoa'],
    default: 'Hoat_dong'
  },
  lanDangNhapCuoi: {
    type: Date,
    default: null
  }
}, {
  timestamps: true,
  collection: 'nguoi_dung'
});

// Index
nguoiDungSchema.index({ email: 1 });
nguoiDungSchema.index({ vaiTro: 1 });

// Hash mật khẩu trước khi lưu
nguoiDungSchema.pre('save', async function(next) {
  if (!this.isModified('matKhauHash')) {
    return next();
  }
  
  try {
    const salt = await bcrypt.genSalt(10);
    this.matKhauHash = await bcrypt.hash(this.matKhauHash, salt);
    next();
  } catch (error) {
    next(error);
  }
});

// Method so sánh mật khẩu
nguoiDungSchema.methods.kiemTraMatKhau = async function(matKhauNhap) {
  return await bcrypt.compare(matKhauNhap, this.matKhauHash);
};

// Ẩn thông tin nhạy cảm khi trả về JSON
nguoiDungSchema.methods.toJSON = function() {
  const obj = this.toObject();
  delete obj.matKhauHash;
  delete obj.__v;
  return obj;
};

const NguoiDung = mongoose.model('NguoiDung', nguoiDungSchema);

module.exports = NguoiDung;